add a new document about how to analyze slow queries (#4035)
pingcap/docs: analyze-slow-queries.md

---
title: Analyze Slow Queries
summary: Learn how to locate and analyze slow queries.
---

# Analyze Slow Queries

To address the issue of slow queries, you need to take the following two steps:

1. Among many queries, identify which type of queries are slow.
2. Analyze why this type of queries are slow.

You can easily perform step 1 using the [slow query log](/identify-slow-queries.md) and the [statement summary table](/statement-summary-tables.md) features. TiDB Dashboard integrates the two features and directly displays the results in your browser in a convenient way. Therefore, this document focuses on step 2.

Generally, slow queries have the following major causes:

- Optimizer issues, such as wrong index selected, wrong join type or sequence selected.
- System issues. All issues not caused by the optimizer are system issues. For example, a busy TiKV instance causes requests to be processed slowly; outdated Region information causes slow queries.

> In actual situations, optimizer issues might cause system issues. For example, in a certain type of queries, if the optimizer that should use index actually uses full table scan, the SQL queries will consume many resources, causing the CPU usage of some KV instances to soar. This might appear to be a system issue, but in essence it is an optimizer issue.

Reviewer: ```suggestion
In actual situations, optimizer issues might cause system issues. For example, for a certain type of queries, the optimizer uses a full table scan instead of the index. As a result, the SQL queries consume many resources, which causes the CPU usage of some TiKV instances to soar. This seems like a system issue, but in essence, it is an optimizer issue.
```

> Identifying system issues is relatively simple. However, to analyze optimizer issues, you need to determine whether the execution plan is reasonable or not. It is recommended to analyze slow queries following these procedures:

Reviewer: ```suggestion
To identify system issues is relatively simple. To analyze optimizer issues, you need to determine whether the execution plan is reasonable or not. Therefore, it is recommended to analyze slow queries by following these procedures:
```

> 1. Identify the query bottleneck, which means the time-consuming part of the query process.

Reviewer: ```suggestion
1. Identify the performance bottleneck of the query, that is, the time-consuming part of the query process.
```

2. Analyze the system issues: analyze the possible causes according to the query bottleneck and the monitoring/log information of that time.
3. Analyze the optimizer issues: analyze whether there is a better execution plan.

The procedures above are explained in the following sections.

> ## Identify the query bottleneck

Reviewer: ```suggestion
## Identify the performance bottleneck of the query
```

> To identify the query bottleneck, you need to have a general understanding of the query process. The key stages of the query execution process in TiDB is illustrated in [TiDB performance map](/media/performance-map.png).

Reviewer: ```suggestion
First, you need to have a general understanding of the query process. The key stages of the query execution process in TiDB are illustrated in [TiDB performance map](/media/performance-map.png).
```

You can get the duration information using the following methods:

- [Slow log](/identify-slow-queries.md). It is recommended to view the slow log in [TiDB Dashboard](/dashboard/dashboard-overview.md).
- [`EXPLAIN ANALYZE` statement](/sql-statements/sql-statement-explain-analyze.md).

The methods above are different in the following aspects:

- The slow log records the duration of almost all stages of a SQL execution, from parsing to returning results, and is relatively comprehensive (you can query and analyze the slow log in TiDB Dashboard in an intuitive way).
- By executing `EXPLAIN ANALYZE`, you can learn the time consumption of each operator in an actual SQL execution. The results have more detailed statistics of the execution duration.

> In summary, you can use the slow log and execute `EXPLAIN ANALYZE` to accurately identify the query bottleneck. The two methods help you determine the SQL query is slow in which component (TiDB or TiKV) at which stage of the execution.

Reviewer: ```suggestion
In summary, the slow log and `EXPLAIN ANALYZE` statements help you determine the SQL query is slow in which component (TiDB or TiKV) at which stage of the execution. Therefore, you can accurately identify the performance bottleneck of the query.
```

In addition, since v4.0.3, the `Plan` field in the slow log also includes the SQL execution information, which is the result of `EXPLAIN ANALYZE`. So you can find all information of SQL duration in the slow log.

## Analyze system issues

> System issues can be divided into the following three types according the execution stage:

Reviewer: ```suggestion
System issues can be divided into the following types according to different execution stages:
```

> 1. TiKV is slow in data processing. For example, the TiKV coprocessor is slow to process data.

Reviewer: ```suggestion
1. TiKV is slow in data processing. For example, the TiKV coprocessor processes data slowly.
```

> 2. TiDB is slow in execution. For example, in the execution stage, a `Join` operator is slow to process data.

Reviewer: ```suggestion
2. TiDB is slow in execution. For example, a `Join` operator processes data slowly.
```
原文中的 "主要指执行阶段" 和 "我们根据执行阶段，分成三个大类" 的“执行阶段”有什么区别？ @qw4990

Author: @qw4990 请帮忙确认一下这两个`执行阶段`有什么区别，谢谢！
![image](https://user-images.githubusercontent.com/41534398/98193889-1fbd9280-1f59-11eb-9800-663fdb02e4c5.png)

Author: @yikeke 已和 @qw4990 确认, "主要指执行阶段"是指 "TiDB 内部执行器执行阶段"，"我们根据执行阶段，分成三个大类"是指"我们根据整个 SQL 的执行阶段"。
```suggestion
2. TiDB is slow in execution. For example, in the execution stage of TiDB's operators, a `Join` operator is slow to process data.
```

Reviewer: ```suggestion
2. TiDB is slow in execution. For example, a `Join` operator processes data slowly.
```
I doubt that "in the execution stage of TiDB's operators" adds any additional information. Consider dropping it to make the sentence concise.

For the description of each field in the slow log, see [fields description](/identify-slow-queries.md#fields-description).

```log
# Query_time: 0.18085
...
# Num_cop_tasks: 1
# Cop_process: Avg_time: 170ms P90_time: 170ms Max_time: 170ms Max_addr: 10.6.131.78
# Cop_wait: Avg_time: 1ms P90_time: 1ms Max_time: 1ms Max_Addr: 10.6.131.78
```

After identifying that TiKV is slow in data processing, you can find out the cause as described in the following sections.

#### TiKV instance is busy

During the execution of a SQL statement, TiDB might fetch data from multiple TiKV instances. If one TiKV instance is slow to respond, the overall SQL execution speed is slowed down.

The `Cop_wait` field in the slow log can help you determine this cause.

```log
# Cop_wait: Avg_time: 1ms P90_time: 2ms Max_time: 110ms Max_Addr: 10.6.131.78
```

From the log above, you can see that a `cop-task` sent to the `10.6.131.78` instance waits `110ms` before it is executed. It might that this instance is busy. You can check the CPU monitoring of that time to confirm the cause.

#### Too many outdated keys

> If a TiKV instance has much outdated data, you need to clean up the data; otherwise, the processing speed is affected.

Author: @yikeke PTAL
```suggestion
A TiKV instance has much outdated data, which needs to be cleaned up for data scan. This impacts the processing speed.
```

> If a TiKV instance has much outdated data, you need to clean up the data; otherwise, the processing speed is affected.

Reviewer: LGTM

Check `Total_keys` and `Processed_keys`. If they are greatly different, the TiKV instance has too many keys of the older versions.

```
...
# Total_keys: 2215187529 Processed_keys: 1108056368
...
```

### Other key stages are slow

#### Slow in getting timestamps

You can compare `Wait_TS` and `Query_time` in the slow log. The timestamps are prefetched, so generally `Wait_TS` should be low.

```
# Query_time: 0.0300000
...
# Wait_TS: 0.02500000
```

#### Outdated Region information

> Region information in the TiDB side might be outdated. In this situation, TiKV might return the `regionMiss` error. Then TiDB gets the Region information from PD again, which is reflected in the `Cop_backoff` information. Both the failed times and the total duration are recorded.

Reviewer: ```suggestion
Region information on the TiDB side might be outdated. In this situation, TiKV might return the `regionMiss` error. Then TiDB gets the Region information from PD again, which is reflected in the `Cop_backoff` information. Both the failed times and the total duration are recorded.
```

---
title: Analyze Slow Queries
summary: Learn how to locate and analyze slow queries.
---

# Analyze Slow Queries

To address the issue of slow queries, you need to take the following two steps:

1. Among many queries, identify which type of queries are slow.
2. Analyze why this type of queries are slow.

> You can easily perform step 1 using the [slow query log](/identify-slow-queries.md) and the [statement summary table](/statement-summary-tables.md) features. It is recommended to use TiDB Dashboard, which integrates the two features and directly displays the slow queries in your browser. 

Reviewer 2: It would be better if we point to the slow query page on TiDB Dashboard: https://docs.pingcap.com/tidb/dev/dashboard-slow-query

Author: Thanks! Updated in https://github.com/pingcap/docs/pull/4035/commits/6f7fd5c7bd076186f18c9a9cdfd3df4667f60301.

> You can easily perform step 1 using the [slow query log](/identify-slow-queries.md) and the [statement summary table](/statement-summary-tables.md) features. It is recommended to use TiDB Dashboard, which integrates the two features and directly displays the slow queries in your browser. 

Author: ```suggestion
You can easily perform step 1 using the [slow query log](/dashboard/dashboard-slow-query.md) and the [statement summary table](/statement-summary-tables.md) features. It is recommended to use [TiDB Dashboard](/dashboard/dashboard-intro.md), which integrates the two features and directly displays the slow queries in your browser. 
```

- [Slow log](/identify-slow-queries.md). It is recommended to view the slow log in [TiDB Dashboard](/dashboard/dashboard-overview.md).
- [`EXPLAIN ANALYZE` statement](/sql-statements/sql-statement-explain-analyze.md).

The methods above are different in the following aspects:

- The slow log records the duration of almost all stages of a SQL execution, from parsing to returning results, and is relatively comprehensive (you can query and analyze the slow log in TiDB Dashboard in an intuitive way).
- By executing `EXPLAIN ANALYZE`, you can learn the time consumption of each operator in an actual SQL execution. The results have more detailed statistics of the execution duration.

In summary, the slow log and `EXPLAIN ANALYZE` statements help you determine the SQL query is slow in which component (TiDB or TiKV) at which stage of the execution. Therefore, you can accurately identify the performance bottleneck of the query.

In addition, since v4.0.3, the `Plan` field in the slow log also includes the SQL execution information, which is the result of `EXPLAIN ANALYZE`. So you can find all information of SQL duration in the slow log.

## Analyze system issues

> System issues can be divided into the following types according to different execution stages:

Author: @yikeke 
```suggestion
System issues can be divided into the following types according to different execution stages of a SQL statement:
```

Reviewer: LGTM

|     └─Selection_58           | 16384.00 |         | cop[tikv] |               | time:0ns, loops:0        | eq(test.t.a, 1)                  | N/A       | N/A  |
|       └─TableFullScan_57     | 16384.00 | -1      | cop[tikv] | table:t       | time:0s, loops:0         | keep order:false                 | N/A       | N/A  |
+------------------------------+----------+---------+-----------+---------------+--------------------------+----------------------------------+-----------+------+
5 rows in set (7.77 sec)
```

But you can identify this type of subquery execution in the slow log:

```
# Query_time: 7.770634843
...
# Rewrite_time: 7.765673663 Preproc_subqueries: 1 Preproc_subqueries_time: 7.765231874
```

From log record above, you can see that a subquery is executed in advance and takes `7.76s`.

### TiDB is slow in execution

Assume that the execution plan in TiDB is correct but the execution is slow. To solve this type of issue, you can adjust parameters or use the hint according to the result of `EXPLAIN ANALYZE` for the SQL statement.

If the execution plan is incorrect, see the [Analyze optimizer issues](#analyze-optimizer-issues) section.

#### Low concurrency

> If you find that the bottleneck is in the operator with concurrency, you can speed up the execution by adjusting the concurrency. For example:

Reviewer: ```suggestion
If the bottleneck is in the operator with concurrency, speed up the execution by adjusting the concurrency. For example:
```

|     └─TableReader_18(Probe)      | 16384.00     | 16384     | root      |               | time:6.880923ms, loops:17                                                           | data:Selection_17                              | 33.6318359375 KB | N/A     |
|       └─Selection_17             | 16384.00     |           | cop[tikv] |               | time:0ns, loops:0                                                                   | not(isnull(test.t.a))                          | N/A              | N/A     |
|         └─TableFullScan_16       | 16384.00     | -1        | cop[tikv] | table:t1      | time:0s, loops:0                                                                    | keep order:false                               | N/A              | N/A     |
+----------------------------------+--------------+-----------+-----------+---------------+-------------------------------------------------------------------------------------+------------------------------------------------+------------------+---------+
9 rows in set (9.67 sec)
```

> For the example above, you can see that `HashJoin_14` and `Projection_24` consume much of the execution time. Consider increasing their concurrency using SQL variables to speed up execution.

Reviewer: ```suggestion
As shown above, `HashJoin_14` and `Projection_24` consume much of the execution time. Consider increasing their concurrency using SQL variables to speed up execution.
```

| HashJoin_8                   | 99800100.00 | root      |               | CARTESIAN inner join, other cond:gt(test.t.a, test.t.a) |
| ├─TableReader_15(Build)      | 9990.00     | root      |               | data:Selection_14                                       |
| │ └─Selection_14             | 9990.00     | cop[tikv] |               | not(isnull(test.t.a))                                   |
| │   └─TableFullScan_13       | 10000.00    | cop[tikv] | table:t2      | keep order:false, stats:pseudo                          |
| └─TableReader_12(Probe)      | 9990.00     | root      |               | data:Selection_11                                       |
|   └─Selection_11             | 9990.00     | cop[tikv] |               | not(isnull(test.t.a))                                   |
|     └─TableFullScan_10       | 10000.00    | cop[tikv] | table:t1      | keep order:false, stats:pseudo                          |
+------------------------------+-------------+-----------+---------------+---------------------------------------------------------+
```

## Analyze optimizer issues

> To analyze optimizer issues, you need to determine whether the execution plan is reasonable or not. You need to have a some understanding of the optimization process and each operator.

Reviewer: ```suggestion
To analyze optimizer issues, you need to determine whether the execution plan is reasonable or not. You need to have some understanding of the optimization process and each operator.
```

> For the following list of examples, the table schema is `create table t (id int, a int, b int, c int, primary key(id), key(a), key(b, c))`.

Reviewer: ```suggestion
For the following examples, assume that the table schema is `create table t (id int, a int, b int, c int, primary key(id), key(a), key(b, c))`.
```
